File: updater/the_judge.go
package updater

import (
	"math"
	"sort"

	"gitlab.com/Njinx/searx-space-autoselector/config"
)

func isOutlier(avgs float64, latency float64, weight float64) bool {
	outlierMultipler := config.ParseConfig().Updater.Advanced.OutlierMultiplier

	if (latency*weight > avgs*outlierMultipler) || (latency < 0) {
		return true
	} else {
		return false
	}
}

func findCanidates(instances *Instances) Canidates {
	conf := config.ParseConfig().Updater.Advanced
	avgs := instances.getTimingAvgs()

	var canidates Canidates
	for _, inst := range instances.instanceList {
		timings := inst.Timings
		if isOutlier(avgs.Initial, timings.Initial, conf.InitialRespWeight) {
			continue
		}
		if isOutlier(avgs.Search, timings.Search, conf.SearchRespWeight) {
			continue
		}
		if isOutlier(avgs.Google, timings.Google, conf.GoogleSearchRespWeight) {
			continue
		}
		if isOutlier(avgs.Wikipedia, timings.Wikipedia, conf.WikipediaSearchRespWeight) {
			continue
		}

		score := timings.Initial/conf.InitialRespWeight + timings.Search/conf.SearchRespWeight + timings.Google/conf.GoogleSearchRespWeight + timings.Wikipedia/conf.WikipediaSearchRespWeight
		score = math.Floor(score*100) / 100

		canidates = append(canidates, Canidate{inst, score})
	}

	sort.Sort(canidates)

	// Now that we've weeded out the bad instances, lets conduct some actual latency
	// tests for more accurate results.
	getUrls := func(x *Canidates) []string {
		var urls []string
		for _, canidate := range canidates {
			urls = append(urls, canidate.Url)
		}

		return urls
	}

	testResults := doLatencyTests(getUrls(&canidates))
	canidates = refineTestCanidates(testResults, &canidates)

	//canidates.reverse() // For use as a stack the best canidates need to be on top
	return canidates
}

func refineTestCanidates(
	testResults []LatencyResponse,
	canidates *Canidates) Canidates {

	resultToCanidate := func(result LatencyResponse, canidates *Canidates) Canidate {
		for _, canidate := range *canidates {
			if canidate.Url == result.addr {
				return canidate
			}
		}

		return Canidate{}
	}

	var newCanidates Canidates
	for _, result := range testResults {
		if result.isAlive {
			/*intensiveResult := doLatencyTestIntensive(result.addr)
			if intensiveResult.isAlive {
				continue
			}*/
			newCanidates = append(newCanidates, resultToCanidate(result, canidates))
		}
	}

	return newCanidates
}

File: the_judge.py
from typing import Coroutine
import utils
from utils import LatencyResponse
import config
import asyncio
import urllib.parse
from common import Canidate, Instance, Instances

def is_outlier(avgs: float, latency: float, weight: float) -> bool:
    if (latency*weight > avgs*config.OUTLIER_MULTIPLIER) or (latency < 0):
        return True
    else:
        return False

def find_canidates(instances: Instances) -> list[Canidate]:
    timing_avgs = instances.get_timing_avgs()
    
    canidates: list[Canidate] = []
    for inst in instances.instance_list:

        if not isinstance(inst.timings.initial, float) \
        or is_outlier(
                timing_avgs.initial,
                inst.timings.initial,
                config.INITIAL_RESP_WEIGHT):
            continue
        if not isinstance(inst.timings.search, float) \
        or is_outlier(
                timing_avgs.search,
                inst.timings.search,
                config.SEARCH_RESP_WEIGHT):
            continue
        if not isinstance(inst.timings.google, float) \
        or is_outlier(
                timing_avgs.google,
                inst.timings.google,
                config.GOOGLE_SEARCH_RESP_WEIGHT):
            continue
        if not isinstance(inst.timings.wikipedia, float) \
        or is_outlier(
                timing_avgs.wikipedia,
                inst.timings.wikipedia,
                config.WIKIPEDIA_SEARCH_RESP_WEIGHT):
            continue

        score = float(inst.timings.initial)   * 1/(config.INITIAL_RESP_WEIGHT) \
              + float(inst.timings.search)    * 1/(config.SEARCH_RESP_WEIGHT) \
              + float(inst.timings.google)    * 1/(config.GOOGLE_SEARCH_RESP_WEIGHT) \
              + float(inst.timings.wikipedia) * 1/(config.WIKIPEDIA_SEARCH_RESP_WEIGHT)
        score = round(score, 2)

        canidates.append(Canidate(inst, score))

    canidates.sort(key=lambda c: c.score)

    # Now that we've weeded out the bad instances, lets conduct some actual latency 
    # tests for more accurate results.
    test_results: list[LatencyResponse] = asyncio.run(utils.do_latency_tests(
        list(map(lambda x: urllib.parse.urlparse(x.instance.url).netloc, canidates))))

    asyncio.run(_refine_test_canidates(test_results, canidates))
    
    canidates.reverse() # For use as a stack; best canidates need to be on top
    return canidates

async def _refine_test_canidates(
        test_results: list[LatencyResponse],
        canidates: list[Canidate]):

    tasks: list[Coroutine] = []
    for result in test_results:
        tasks.append(_refine_test_canidates_iter(result))

    for ret in await asyncio.gather(*(tasks)):
        if not ret[0]:
            test_result: LatencyResponse = ret[1]
            canidates[:] = [c for c in canidates if c.instance.url != test_result.addr]

async def _refine_test_canidates_iter(result: LatencyResponse) \
        -> tuple[LatencyResponse, bool]:
    if not result.is_alive:
        res_hostname: str = urllib.parse.urlparse(result.addr).netloc
        intensive_result: LatencyResponse = await utils.do_latency_test_intensive(res_hostname)
        if not intensive_result.is_alive:
            return (result, False)

    return (result, True)
